File: chapter-2/2.1.js
/*
Write code to remove duplicates from an unsorted linked list. 
FOLLOW UP How would you solve this problem if a temporary buffer is not allowed?
*/

// Complexity O(N)
var removeDuplicates = function(node) {
  var duplicates = {};
  duplicates[node.data] = true;
  var current = node.next;

  while(current != null) {
    if (current.data in duplicates) {
      var prevNode = current.prev;
      var nextNode = current.next;
      prevNode.next = current.next;
      nextNode.prev = prevNode
    } else {
      duplicates[current.data] = true;
    }

    current = current.next;
  }
};

File: chapter-2/ruby/2.1.rb
# Implement an algorithm to find the nth to last element of a singly linked list.

# Complexity O(N)
def remove_duplacate_nodes node
  unless node[:node_next]
    return
  end

  duplicates = []
  duplicates.push node[:data]

  current = node[:node_next]

  while current do
    if duplicates.include? current[:data]
      current[:node_prev][:node_next] = current[:node_next]
      current[:node_next][:node_prev] = current[:node_prev]
    else
      duplicates.push current[:data]
    end

    current = current[:node_next]
  end
end
